### PR TITLE
Adding more documentation for upsampling with replacement and error m…

### LINKED ISSUE #27451: `Dataframe.sample()` isn't validating `frac`
Minor/trivial bug:

https://github.com/pandas-dev/pandas/blob/master/pandas/core/generic.py#L4960

Shouldn't `sample()` check that `frac` is between 0 and 1 somewhere?

Doc improvement suggestion:

Could anyone confirm that `replace=True|False` means allowing|disallowing sampling the same row more than once?

Methinks the docs could use an update to make what `replace` does clearer for newcomers - or at least for those who aren't familiar with `numpy`.

### PR DESCRIPTION
**Pull request by CooperData**
…essage in case replacement is set to False

- [X] closes #27451
- [x] tests added / passed
- [X] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/generic.py [hunk 1]
```diff
@@ -4964,6 +4964,21 @@ def sample(
         dog          4          0                  2
         fish         0          0                  8
 
+
+        An upsample sample of the ``DataFrame`` with replacement:
+        Note that `replace` parameter has to be `True` for `frac` parameter > 1.
```
Comment (jreback) [inline pandas/core/generic.py]: do we have this in the Notes section as well (about replace=True for frac > 1)

File: pandas/tests/generic/test_generic.py [hunk 2]
```diff
@@ -424,6 +434,14 @@ def test_sample(self):
         weights_with_None[5] = 0.5
         self._compare(o.sample(n=1, axis=0, weights=weights_with_None), o.iloc[5:6])
 
+    def test_sample_upsampling_without_replacement(self):
+        df = pd.DataFrame({"A": list("abc")})
+        msg = (
+            "Replace has to be set to `True` when upsampling the population `frac` > 1"
+        )
+        with pytest.raises(TypeError, match=msg):
+            hash(df.sample(frac=2, replace=False))
```
Comment (jreback) [inline pandas/tests/generic/test_generic.py]: don't need the hash

File: pandas/tests/generic/test_generic.py [hunk 3]
```diff
@@ -424,6 +434,14 @@ def test_sample(self):
         weights_with_None[5] = 0.5
         self._compare(o.sample(n=1, axis=0, weights=weights_with_None), o.iloc[5:6])
 
+    def test_sample_upsampling_without_replacement(self):
+        df = pd.DataFrame({"A": list("abc")})
```
Comment (jreback) [inline pandas/tests/generic/test_generic.py]: can you add the issue number as a comment

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — CHANGES_REQUESTED** [review]
go ahead and add a whatsnew note, bugfix in 1.0 in the numeric section; you can say that a better error message will be raised when frac > 1 and replace=True

File: pandas/core/generic.py (hunk 1 above)
Comment (CooperData) [reply pandas/core/generic.py]: I added the Notes session

File: pandas/tests/generic/test_generic.py (hunk 3 above)
Comment (CooperData) [reply pandas/tests/generic/test_generic.py]: I removed the hash

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. ping on green.

**Comment by CooperData** [comment]
@jreback - CI failed but I am not sure what it's causing the issue. I was able to run the tests while I was at the sprint. But dont know if I did anything that is creating the issue.

**Comment by jreback** [comment]
see https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=20307&view=logs&jobId=d6a36677-9147-59f9-684d-001fc5c44c5e

looks like just some extra white space 

fix and push up if u can

**Comment by CooperData** [comment]
I added a dot at the end of the raised error as per other errors I saw. I also changed the test msg checked to match the error raised so it's 2 lines as well. I was not able to understand what exactly the error was.

**Comment by jreback** [comment]
thanks @CooperData

**Comment by CooperData** [comment]
@jreback, I am going through the messages on the build for the last push, and it seems it's failing because the function inside that is supposed to give the error message is failing. In other words, the desired behavior is to make the test fail but compare the error message, not to make the test fail itself.

**Comment by CooperData** [comment]
The failures seem associated with something else that I did not change: [](https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=20397&view=logs&jobId=5ddd427e-1f95-5998-4195-7ffbeb9b114d&taskId=607906cd-31fb-52fc-bb5f-740f55a1afa6&lineStart=150&lineEnd=186&colStart=1&colEnd=126)

**Comment by jreback** [comment]
thanks @CooperData !

**Pull request merged**